packages/react-scheduler/src/ReactScheduler.js

```diff
@@ -100,14 +128,86 @@ if (!ExecutionEnvironment.canUseDOM) {
   let previousFrameTime = 33;
   let activeFrameTime = 33;
 
-  const frameDeadlineObject = {
+  const frameDeadlineObject: Deadline = {
     didTimeout: false,
     timeRemaining() {
       const remaining = frameDeadline - now();
       return remaining > 0 ? remaining : 0;
     },
   };
 
+  const safelyCallScheduledCallback = function(callback, callbackId) {
+    if (!registeredCallbackIds[callbackId]) {
+      // ignore cancelled callbacks
+      return;
+    }
+    try {
+      callback(frameDeadlineObject);
+      // Avoid using 'catch' to keep errors easy to debug
+    } finally {
+      // always clean up the callbackId, even if the callback throws
+      delete registeredCallbackIds[callbackId];
+    }
+  };
+
+  /**
+   * Checks for timed out callbacks, runs them, and then checks again to see if
+   * any more have timed out.
+   * Keeps doing this until there are none which have currently timed out.
+   */
+  const callTimedOutCallbacks = function() {
+    if (pendingCallbacks.length === 0) {
+      return;
+    }
+
+    let currentTime = now();
+    // TODO: this would be more efficient if deferred callbacks are stored in
+    // min heap.
+    // Or in a linked list with links for both timeoutTime order and insertion
+    // order.
+    // For now an easy compromise is the current approach:
+    // Keep a pointer to the soonest timeoutTime, and check that first.
+    // If it has not expired, we can skip traversing the whole list.
+    // If it has expired, then we step through all the callbacks.
+    if (nextSoonestTimeoutTime === -1 || nextSoonestTimeoutTime > currentTime) {
+      // We know that none of them have timed out yet.
+      return;
+    }
+    nextSoonestTimeoutTime = -1; // we will reset it below
+
+    // keep checking until we don't find any more timed out callbacks
+    let foundTimedOutCallback = false;
+    frameDeadlineObject.didTimeout = true;
+    do {
+      currentTime = now();
+      foundTimedOutCallback = false;
+      for (let i = 0, len = pendingCallbacks.length; i < len; i++) {
+        const currentCallbackConfig = pendingCallbacks[i];
+        if (!currentCallbackConfig) {
+          // we might go off the end of the array
+          continue;
+        }
+        const timeoutTime = currentCallbackConfig.timeoutTime;
+        if (timeoutTime !== -1 && timeoutTime <= currentTime) {
+          // it has timed out!
+          foundTimedOutCallback = true;
```

Comment:
What is this for? Don't we have to check the entire list regardless?

I don't understand why this is a nested loop. What is the purpose of the outer loop?

Reply:
Discussing in person, we agreed that it could just get pushed into the next tick if something times out while one cb is running.